### PR TITLE
Implement CUBE LUT domain properly, tetrahedral interpolation for 3D LUTs

### PR DESCRIPTION
**Pull request by jpark37**
### Description
LUT improvements:

- Previously assumed zero to one domain instead of respecting values.
- Tetrahedral interpolation for 3D LUTs instead of trilinear.

We have separate shaders for 1D and 3D LUTs now because domain handling is per-channel for 1D LUTs, and all channels combined for 3D LUTs. We also need separate shaders for tetrahedral interpolation for 3D LUTs as requested by the CUBE spec. 1D LUTs are still linearly interpolated per channel.

Tetrahedral interpolation math source: https://www.nvidia.com/content/GTC/posters/2010/V01-Real-Time-Color-Space-Conversion-for-High-Resolution-Video.pdf

### Motivation and Context
If anyone uses DOMAIN_MIN and DOMAIN_MAX in their CUBE file, they will have wrong colors.

If anyone expected tetrahedral interpolation, they were not getting it.

### How Has This Been Tested?
PNG, 1D, and 3D LUTs still work.

Tried various values for domains, and inspected math using RenderDoc shader debugger. Checked switching between PNG and CUBE files still worked without bleeding internal state. Checked invalid domains (non-positive intervals) are rejected and logged for R, G, and B channels of 1D and 3D CUBE LUTs. CUBE LUTs without domain specifiers are still handled correctly, and default to [0-1].

Tried wacky LUT to expose difference between trilinear and tetrahedral interpolation. Used RenderDoc shader debugger to verify tetrahedron selection was working.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

**Comment by jpark37** [comment]
I'm sort of cheating bolting on the extra commit, but it builds on top of the first. I can split the PR into two later if you want to approve the former commit without the latter.

File: plugins/obs-filters/color-grade-filter.c [hunk 1]
```diff
@@ -292,16 +286,46 @@ static void color_grade_filter_update(void *data, obs_data_t *settings)
...
 			const uint32_t width = filter->cube_width;
-			filter->target = gs_voltexture_create(
-				width, width, width, GS_RGBA16F, 1,
-				(uint8_t **)&filter->cube_data, 0);
-			filter->clut_scale = (float)(width - 1) / (float)width;
-			filter->clut_offset = 0.5f / (float)width;
+			if (filter->clut_dim == CLUT_1D) {
+				filter->target = gs_texture_create(
+					width, 1, GS_RGBA16F, 1,
+					(uint8_t **)&filter->cube_data, 0);
```
Comment (kkartaltepe) [inline plugins/obs-filters/color-grade-filter.c:298]: Use correct const-ness to resolve compiler warnings.
(const uint8_t **)&filter->cube_data

Comment (jpark37) [reply plugins/obs-filters/color-grade-filter.c]: Having to do this extra const cast seems like a silly oversight in the C spec that they should just copy over from C++. Added.

Comment (kkartaltepe) [reply plugins/obs-filters/color-grade-filter.c]: I agree, and im sure they'll fix it in C20 :)

Comment (c3r1c3) [reply plugins/obs-filters/color-grade-filter.c]: Yeah, but MS won't implement it so you'll be stuck doing this on Windows forever. ;-)

**Pull request merged**